### PR TITLE
Added the option to specify custom dimensions for a Matomo tag

### LINKED ISSUE #806: [Enhancement] Add Custom Dimension tracking as an option for Matomo Analytics Tags
The only way to track Custom Dimensions on demand via the Matomo Tag Manager is to use a Custom HTML Tag. The default Custom Dimension field in the Matomo Configuration Variable is not suitable/practical for setting trigger-based Custom Dimensions.

We should be able to add Custom Dimension tracking as a 'Tracking Type' option for Matomo Analytics Tags. For example...
image

The options for this tracking type would be Custom Dimension ID and value - where the value can accept a variable.

### PR DESCRIPTION
**Pull request by snake14**
### Description:

Added the option to specify custom dimensions for a Matomo tag.

Fixes: #806 

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: Updates/5.2.0-b4.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+use Piwik\Plugins\TagManager\UpdateHelper\NewTagParameterMigrator;
+use Piwik\Updater;
+use Piwik\Updater\Migration;
+use Piwik\Updater\Migration\Factory as MigrationFactory;
+use Piwik\Updates as PiwikUpdates;
+
+/**
+ * Update for version 5.2.0-b4.
+ */
+class Updates_5_2_0_b4 extends PiwikUpdates
```
Comment (AltamashShaikh) [inline Updates/5.2.0-b4.php:22]: Note: Update submodules after you merge this PR @snake14

File: Template/Tag/MatomoTag.php [hunk 2]
```diff
@@ -205,7 +205,46 @@ public function getParameters()
...
+
+                $field->uiControl = FieldConfig::UI_CONTROL_MULTI_TUPLE;
+                $field1 = new FieldConfig\MultiPair('Index', 'index', FieldConfig::UI_CONTROL_TEXT);
+                $field1->customFieldComponent = self::FIELD_VARIABLE_COMPONENT;
+                $field2 = new FieldConfig\MultiPair('Value', 'value', FieldConfig::UI_CONTROL_TEXT);
+                $field2->customFieldComponent = self::FIELD_VARIABLE_COMPONENT;
+                $field->uiControlAttributes['field1'] = $field1->toArray();
+                $field->uiControlAttributes['field2'] = $field2->toArray();
+            }),
+            $this->makeSetting('areCustomDimensionsSticky', false, FieldConfig::TYPE_BOOL, function (FieldConfig $field) {
```
Comment (AltamashShaikh) [inline Template/Tag/MatomoTag.php:243]: @snake14 Without setting this config, the dimensions are still persisting

Screenshot from 2024-09-27 11-32-00



Screenshot from 2024-09-27 11-35-03

**Review by AltamashShaikh — COMMENTED** [review]
@snake14 Left 1 comment, based on that I can test further

File: Template/Tag/MatomoTag.php (hunk 2 above)
Comment (snake14) [reply Template/Tag/MatomoTag.php:243]: @AltamashShaikh I'm unable to reproduce. When I have the checkbox disabled, the custom dimension is only used when the specific tag is triggered. For example, I have the custom dimension configured for a tag using an image click trigger. When I click other elements, the dimension isn't present in the request, clicking the image includes the dimension, and the dimension is gone again when I click a different element.

Comment (snake14) [reply Template/Tag/MatomoTag.php:243]: @AltamashShaikh I just recorded my demo video (see Slack channel) and it worked as expected. I also tested on demo2 and it worked correctly.

Comment (AltamashShaikh) [reply Template/Tag/MatomoTag.php:243]: @snake14 If it works we can merge it

**Review by AltamashShaikh — APPROVED** [review]

**Pull request merged**